File: src/main/java/com/hazza/LeetCodeByJava/No2AddTwoNumbers.java
package com.hazza.LeetCodeByJava;

/**
 * https://leetcode.com/problems/add-two-numbers/#/description
 */
public class No2AddTwoNumbers {
    private class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(0);
        ListNode tmp = res;
        int sum = 0;

        while(l1 != null || l2 != null) {
            sum /= 10;  //上一次循环留下的sum值
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }
            tmp.next = new ListNode(sum % 10);
            tmp = tmp.next;
        }
        if (sum / 10 == 1) {
            tmp.next = new ListNode(1);
        }
        return res.next;
    }
}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No2_Add_Two_Numbers.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/2 19:16
# https://leetcode.com/problems/add-two-numbers/

from ListNode import ListNode, makeListNode, printListNode


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        res = ListNode(0)
        tmp = res
        left = 0

        while l1 or l2 or left:
            left, right = divmod(sum(l and l.val or 0 for l in (l1, l2)) + left, 10)
            tmp.next = ListNode(right)
            tmp = tmp.next
            l1 = l1 and l1.next
            l2 = l2 and l2.next

        return res.next


if __name__ == '__main__':
    solution = Solution()
    l1 = makeListNode([2, 4, 3])
    l2 = makeListNode([5, 6, 4])
    res = solution.addTwoNumbers(l1, l2)
    printListNode(res)
